README.md

# Initial commit

Repository transfer of a very very old project,
with the hope of being useful for someone with
interest for operating systems making.



heavenos-0.1.4/CHANGELOG
0.1.4
FIRST RELEASE

heavenos-0.1.4/test_kernel.asm
[BITS 16]
[ORG 0x00000500]

mov ax, 0xB800
mov es, ax

;;; RM_Clear_Screen
mov cx, 0x0fa0
mov di, 0
next_char0:
mov dword [es:di], 0
add di, 4
cmp di, cx
jne next_char0

;;; RM_Write_Message
mov ax, 0xB800
mov es, ax
;; RM_Write_Boot_Mensage
mov di, 160*0x02 + 2*0x01
mov si, kernel_msg
mov ah, 0x03
next_char1:
mov al, byte [si]
mov word [es:di], ax
times 2 inc di
inc si
cmp byte [si],0
jne next_char1

check_keyboard:
in   al, 0x60                
cmp  al, 0x01     	; check ESC key           
jne  check_keyboard 



kernel_msg	 	DB 'Loading HeavenOS Kernel V0.1.4',0



heavenos-0.1.4/bootloaders/boot_test.asm
[BITS 16]
[ORG 0x7C00]

; DOS header alike boot sector 
;==========================================================
;DB 0xEB,0x1E,0x90 ; SHORT JUMP to 0x1E
;DB 'HeavenOS'     ; OEM Name and Number
;DW 512		  ; Bytes per Sector (512)
;DB 0	 	  ; Sectors per Allocation Unit (cluster)
;DW 0		  ; Reserved seectors (for Boot Record)
;DB 0		  ; Number of FATs
;DW 0		  ; Number of Root Directories Entries
;DW 0		  ; Number of logical Sectors
;DB 0xF0           ; Medium Descriptor Byte (0xF0 - FD, 0xF8 - HD)   
;DW 0		  ; Sectors per FAT
;DW 18		  ; Sectors per TRACK
;DW 2		  ; Number of Heads
;DW 0		  ; Number of Hidden Sectors
;DW 0xAA55	  ; Load Signature - 0xAA55	
;==========================================================
; Boot Loader
;==========================================================

; save EDX value in memory (the last 4 bytes of conventional memory)
; CPU ID Information
mov ax, 0x9000
mov es, ax
mov di, 0xFFFC
mov dword [es:di], edx

; setting regs
mov ax, 0 
mov ds, ax
mov es, ax
mov ss, ax
;mov sp, (0x7C00 + 0x300)
;mov bp, 0 

jmp  0:start

; COLD REBOOT
cold_boot:
mov             ax, 0x0040                      ; ax = 0040h
mov             es, ax                          ; point es to ax
mov             word [es:0072h], 0x0000         ; Reset
db              0xEA                            ; JUMP
dw              0x0000                          ; FFFF:0000
dw              0xFFFF                          ; ! REBOOTING !
;int     0x19                                   ; warm boot pc 

start:

;=======================================================================
; Check if CPU is 386+
;=======================================================================
pushf                   ; save flags for later
xor     ah,ah           ; clear high byte
push    ax              ; push AX on the stack
popf                    ; pop this value into the flag register
pushf                   ; push flags on the stack
pop     ax              ; ...and get flags into AX
and     ah,0f0h         ; try to set the high nibble
cmp     ah,0f0h         ; on a 80386, the high nibble can never be 0f0h
je      cold_boot
mov     ah,70h          ; now try to set NT and IOPL
push    ax
popf
pushf
pop     ax
and     ah,70h          ; if they couldn't be modified, no 386 is installed
jz      cold_boot
popf                    ; restore flags
;=======================================================================

; set ES to video address
mov ax, 0xB800
mov es, ax

;;; RM_Clear_Screen
mov cx, 0x0fa0
mov di, 0
next_char0:
mov dword [es:di], 0
add di, 4
cmp di, cx
jne next_char0

;;; RM_Write_Boot_Mensage
mov di, 160*0x01 + 2*0x01
mov si, boot_msg
mov ah, 0x03
next_char:
mov al, byte [si]
mov word [es:di], ax
times 2 inc di
inc si
cmp byte [si],0
jne next_char

; BIOS disk reads should be retried at least three times and the
; controller should be reset upon error detection
; be sure ES:BX does not cross a 64K segment boundary or a
; DMA boundary error will occur
; many programming references list only floppy disk register values
; only the disk number is checked for validity
; the parameters in CX change depending on the number of cylinders;
; the track/cylinder number is a 10 bit value taken from the 2 high
; order bits of CL and the 8 bits in CH (low order 8 bits of track):

;=======================================================================
; Copying Kernel Image from Floppy Filesystem to Memory
; point es:bx to 0x0050:0x0000
;=======================================================================
copy_kernel:
mov    ax, 0x0050  		; segment
mov    es, ax          
mov    bx, 0x0000  		; offset 
; load data from disk to [es:bx]
mov    ah, 0x02    		; read sector(s)
mov    dl, 0x00    		; disk (0x00 - First FD, 0x80 - First HD)
mov    dh, 0       		; head  
mov    ch, 0       		; track (0-1023)
mov    cl, 2      			; sector (2 higher bits = 10-9 bits of track)
mov    al, 17	   		; max. sectors/head = 18, first sector = bootloader
int    0x13


;cmp    ah, 0
;je     skip_reset
;do_reset:
;mov    ah, 0x00    		; reset drive
;mov    dl, 0x00    		; disk type 
;xor    ax, ax
;int    0x13
;jmp short copy_kernel
;skip_reset:


;====================================================================
; TODO : Read a Sector for Protection in the future
; Password protection ?
;====================================================================

mov ax, 0xB800
mov es, ax
;;; RM_Write_Boot_Mensage
mov di, 160*0x02 + 2*0x01
mov si, jump_msg
mov ah, 0x03
next_char1:
mov al, byte [si]
mov word [es:di], ax
times 2 inc di
inc si
cmp byte [si],0
jne next_char1

check_keyboard:
in   al, 0x60                
cmp  al, 0x01     	; check ESC key           
jne  check_keyboard          
;mov ah,0	; await key pressed
;int 16h

; Disable Floppy Motor
;mov al,0001100b  ;Disable all motors, controller enabled, DMA  & IRQ enabled
;mov dx,0x3F2     ;Digital Output Register
;out dx,al        ;Sending the command to the controller

; turn off floppy motor
mov dx,3F2h
xor al,al
out dx,al

; JUMP to address 0050:0000 RM (0x00000500 PM)
jmp  0x0050:0


;=======================================================================
boot_msg		DB 	'Loading Boot Sector...',0 
jump_msg		DB	'Press ESC for Jumping to Kernel 32-bit PM Code...',0

mov    ax, 0x0050  		; segment
mov    es, ax          
mov    bx, 0x0000  		; offset 
; load data from disk to [es:bx]
mov    ah, 0x02    		; read sector(s)
mov    dl, 0x00    		; disk (0x00 - First FD, 0x80 - First HD)
mov    dh, 0       		; head  
mov    ch, 0       		; track (0-1023)
mov    cl, 2      			; sector (2 higher bits = 10-9 bits of track)
mov    al, 17	   		; max. sectors = 18 
int  	0x13
add	bx, 512*17
mov dh, 1			; another head  
mov ch, 0			; track 0 (0-1023)
next_sectors:
mov    ax, 0x0050  		; segment
mov    es, ax          
;mov    bx, 0x0000  		; offset 
; load data from disk to [es:bx]
mov    ah, 0x02    		; read sector(s)
mov    dl, 0x00    		; disk (0x00 - First FD, 0x80 - First HD)
;mov    dh, 0       		; head  
;mov    ch, 0       		; track (0-1023)
mov    cl, 1      			; sector (2 higher bits = 10-9 bit of track)
mov    al, 18	   		; max. sectors = 18 

pusha
int    0x13
popa

inc	dh
add	bx, 512*18
cmp dh, 2
jne next_head
mov	dh, 0
inc ch
next_head:
cmp ch, 2
jne next_sectors 


;S = (lsn mod spt) + 1
;H = (lsn / spt) mod nh
;T = lsn / (spt*nh)
;
;S is the physical sector number
;H is the head number
;T is the track number
;lsn is the logical sector number
;spt is the number of sectors per track
;nh is the number of heads 
;/ is an integer division
;mod is the modulo operation

;-----------------------------------------------------------------------
times 510-($-$$) DB 0
DW	0xAA55 		; BOOT LOADER SIGNATURE
;-----------------------------------------------------------------------

; PM Address(REAL) = 16 * RM Segment + RM Offset
;  Memory Address   
;     65536        = 16 * 4096 + 0 
;                  = (RM Segment << 4) + RM Offset
;
; 16*0x1000 = 65536*1 start first segment 
; 16*0x2000 = 65536*2 start second segment
; 16*0x3000 = 65536*3 start third segment


heavenos-0.1.4/kernel.asm
;**********************************************************************
; Copyright 2006, Nuno Miguel (codingheaven@gmail.com) 
; All rights reserved. 
;
; See the LICENSE file
;
; The origin of this software must not be misrepresented, either by
; explicit claim or by omission.  Since few users ever read sources,
; credits must appear in the documentation.
;
; CODE IMPROVEMENTS by:
;	Nuno Miguel (codingheaven@gmail.com)	
;
;       
;
;
;
;***********************************************************************


;=============================================================;
; HeavenOS Micro-Kernel V0.1.4
; In the future, Version Number should be only on VERSION file 							               ;
;=============================================================;
;    500 - 9FFFF Free Memory < 1MB (Conventional 640K)                        ;
; 100000 - ?     Free Memory >= 1MB (Extended)                                ;
;-----------------------------------------------------------------------------;

[BITS 16]
;[ORG 0x00000500]
;[ORG 0x00010000]
[ORG 0x00001000]
;-----------------------------------------------------------------------------;
; 16 BITS CODE                                                                ;
;-----------------------------------------------------------------------------;

; set ES to video address
mov ax, 0xB800
mov es, ax

;;; RM_Clear_Screen
mov cx, 0x0fa0
mov di, 0
next_char0:
mov dword [es:di], 0
add di, 4
cmp di, cx
jne next_char0

;;; RM_Write_Message
mov ax, 0xB800
mov es, ax
;; RM_Write_Boot_Mensage
mov di, 160*0x02 + 2*0x01
mov si, kernel_msg
mov ah, 0x03
next_char1:
mov al, byte [si]
mov word [es:di], ax
times 2 inc di
inc si
cmp byte [si],0
jne next_char1

check_keyboard:
in   al, 0x60                
cmp  al, 0x1C    	; check ENTER key           
jne  check_keyboard      

;---------------------------------------------------------------------------;
; INITIALIZATION STAGE FOR THE 32-BIT PM SWITCH                             ;
;---------------------------------------------------------------------------;
cli     ; disable interrupts

; Enable A20 on PS/2 compatibles
;in      al,0x92
;or      al,2
;out     0x92,al

; Enable A20 on AT compatibles > 1MB
mov     al,0xD1
out     0x64,al
mov     al,0xDF
out     0x60,al

lgdt [gdt_ptr]			; Load GDT into GDTR

; switch to PM
mov     eax, cr0                ; Get Control Register 0
or      eax, 0x01               ; set PE bit (bit 0) in (e)ax
mov     cr0, eax                ; Switch to Protected Mode

; setting regs
mov ax, 0x0010
mov es, ax      ; extra segment
mov ds, ax      ; data segment
mov ss, ax      ; stack segment -> grows down in memory
mov fs, ax
mov gs, ax
mov esp, 0x5000         ; stack pointer
mov ebp, 0x5000         ; base pointer

jmp dword 0x0008:kernel_start 

;---------------------------------------------------------------------------;
; 32-BIT CODE                                                              ;
;---------------------------------------------------------------------------;
[BITS 32]

%include "cpu.inc"

%include "util.inc"
%include "dma.inc"
%include "pic.inc"
%include "floppy.inc"
%include "pit.inc"
%include "cmos.inc"
%include "rtc.inc"
%include "timer.inc"
%include "video.inc"
%include "keyboard.inc"

%include "gdt.inc"
%include "idt.inc";

%include "memory.inc"
%include "bios.inc"
%include "vesa.inc"
%include "pci.inc"
%include "isapnp.inc"
;%include "shellcmds.inc"
%include "tss.inc"

%include "filesystem.inc"


;---------------------------------------------------------------------------;
; KERNEL MAIN CODE                                                          ;
;---------------------------------------------------------------------------;
kernel_start:

call dword 0x0008:fill_idt	; FILL IDT with INT Addresses
lidt [idt_ptr]                  ; Load IDT into IDTR

call dword 0x0008:PIC_setup	; SETUP PIC

PIC_master_set_irqs     10111001b ; Disable IRQ0 (Timer INT) 

call dword 0x0008:PIT_setup	; SETUP PIT

KEYBOARD_init			; Initialize Keyboard

RTC_int_enable                  ; Enable RTC Interrupts

sti				; Enable Interrupts

; TESTING DEBUG ON EXCEPTION (REAL COOL! =)
;dd 0xABCD00DA			; this will run Exception 13	(GENERAL PROTECTION)

VIDEO_TEXTMODE_write_mensage 		0x03,0x01,0x07,kernel_msg
KEYBOARD_check_key 0x1C
;===============================================;
; Copying Filesystem from Floppy to Ramdisk...  ;
;===============================================;
;VIDEO_TEXTMODE_write_mensage 		0x04,0x01,0x02,copy_filesystem_msg
;FLOPPY_copy_to_address

;RAMDISK_read
;VIDEO_TEXTMODE_write_mensage            0x05,0x01,0x07,RAMDISK_BUFFER
;===============================================;
; Begin Multi-Tasking!
;===============================================;
TIMER_rtc_msdelay	2000
call dword 0x0008:VIDEO_VGA_fadeout
call dword 0x0008:VIDEO_TEXTMODE_clear_screen

call dword 0x0008:init_tss	; Initialize TSS's

mov ax, sys_tss-gdt 
ltr ax				; Load First TSS Descriptor into TR
cli
PIC_master_set_irqs     10111000b ; Enable IRQ0 (Timer INT) 
jmp dword 0x0008:sheduler

;=============================================================================;
kernel_msg	 	DB 'Loading HeavenOS Kernel V0.1.4',0
dword_number		DD 0
dword_number2		DD 0
word_number		DW 0
bytenumber		DB 0
number_string		DB '          ',0
copy_filesystem_msg	DB 'Copying Filesystem from Floppy to Ramdisk...',0
temp_string		DB '                                                                               ',0
;=============================================================================;
;times 64000-($-$$) 	DB 0
;640x1024 = max_kernel_size



heavenos-0.1.4/bootloaders/boot.asm
;
; By Nuno Miguel (codingheaven@gmail.com)
;
[BITS 16]
[ORG 0x7C00]

; DOS header alike boot sector 
;==========================================================
;DB 0xEB,0x1E,0x90 ; SHORT JUMP to 0x1E
;DB 'HeavenOS'     ; OEM Name and Number
;DW 512		  ; Bytes per Sector (512)
;DB 0	 	  ; Sectors per Allocation Unit (cluster)
;DW 0		  ; Reserved seectors (for Boot Record)
;DB 0		  ; Number of FATs
;DW 0		  ; Number of Root Directories Entries
;DW 0		  ; Number of logical Sectors
;DB 0xF0           ; Medium Descriptor Byte (0xF0 - FD, 0xF8 - HD)   
;DW 0		  ; Sectors per FAT
;DW 18		  ; Sectors per TRACK
;DW 2		  ; Number of Heads
;DW 0		  ; Number of Hidden Sectors
;DW 0xAA55	  ; Load Signature - 0xAA55	
;==========================================================
; Boot Loader
;==========================================================

; save EDX value in memory (the last 4 bytes of conventional memory)
; CPU ID Information
mov ax, 0x9000
mov es, ax
mov di, 0xFFFC
mov dword [es:di], edx

; setting regs
mov ax, 0 
mov ds, ax
mov es, ax
mov ss, ax
mov sp, (0x7C00 + 0x400)
mov bp, 0 

jmp  0:start

%macro print_number     3
mov ax, 0xB800
mov es, ax
mov di, 160*%1 + 2*(%2+5)
mov ax, %3
mov bx, 10
%%next_digit:
div bx
mov dh, 0x03
add dl, 48
mov word [es:di], dx
times 2 dec di
cmp ax, 10
jge %%next_digit
mov word [es:di], ax
%endmacro

; COLD REBOOT
cold_boot:
mov             ax, 0x0040                      ; ax = 0040h
mov             es, ax                          ; point es to ax
mov             word [es:0072h], 0x0000         ; Reset
db              0xEA                            ; JUMP
dw              0x0000                          ; FFFF:0000
dw              0xFFFF                          ; ! REBOOTING !
;int     0x19                                   ; warm boot pc 

start:

;=======================================================================
; Check if CPU is 386+
;=======================================================================
pushf                   ; save flags for later
xor     ah,ah           ; clear high byte
push    ax              ; push AX on the stack
popf                    ; pop this value into the flag register
pushf                   ; push flags on the stack
pop     ax              ; ...and get flags into AX
and     ah,0f0h         ; try to set the high nibble
cmp     ah,0f0h         ; on a 80386, the high nibble can never be 0f0h
je      cold_boot
mov     ah,70h          ; now try to set NT and IOPL
push    ax
popf
pushf
pop     ax
and     ah,70h          ; if they couldn't be modified, no 386 is installed
jz      cold_boot
popf                    ; restore flags
;=======================================================================

; set ES to video address
mov ax, 0xB800
mov es, ax

;;; RM_Clear_Screen
mov cx, 0x0fa0
mov di, 0
next_char0:
mov dword [es:di], 0
add di, 4
cmp di, cx
jne next_char0

;;; RM_Write_Boot_Mensage
mov di, 160*0x01 + 2*0x01
mov si, boot_msg
mov ah, 0x03
next_char:
mov al, byte [si]
mov word [es:di], ax
times 2 inc di
inc si
cmp byte [si],0
jne next_char


;=======================================================================
; Checking Kernel Image Length
;=======================================================================
mov    ax, 0x0050
mov    es, ax
mov    bx, 0x0000
mov    ah, 0x02
mov    dl, 0x00
mov    dh, HEADER_HEAD
mov    ch, HEADER_TRACK
mov    cl, HEADER_SECTOR
mov    al, 19
sub    al, cl
int    0x13
mov    ax, word [es:bx+36]
mov    [KERNEL_SIZE], ax
;=======================================================================

;mov    word [KERNEL_SIZE], 5000
;cmp     ax, 28230
;jne    skip_msg

;mov ax, 0xB800
;mov es, ax
;mov di, 160*0x03 + 2*0x01
;mov si, boot_msg
;mov ah, 0x03
;next_char1:
;mov al, byte [si]
;mov word [es:di], ax
;times 2 inc di
;inc si
;cmp byte [si],0
;jne next_char1
;skip_msg:


;=======================================================================
; Copying Kernel Image from Floppy Filesystem to Memory
; point es:bx to 0x0050:0x0000
;=======================================================================
mov    ax, 0x0050  		; segment
mov    es, ax          
mov    bx, 0x0000  		; offset 
; load data from disk to [es:bx]
mov    ah, 0x02    		; read sector(s)
mov    dl, 0x00    		; disk (0x00 - First FD, 0x80 - First HD)
mov    dh, KERNEL_HEAD       	; head  
mov    ch, KERNEL_TRACK       	; track (0-1023)
mov    cl, KERNEL_SECTOR      	; sector (2 higher bits = 10-9 bits of track)

mov    al, 19	   		; max. sectors = 18 
sub    al, cl	   		; n sectors
mov    [ADDRESS_POSITION], bx
pusha
int    0x13
xor    ah, ah 
mov    bx, 512
mul    bx
add    [ADDRESS_POSITION], ax
next_head:
popa
inc    dh
cmp    dh, 0x02    		; head < 2 ?
jne     skip_head
mov    dh, 0			; head 0
inc    ch   			; next track
skip_head:
mov    ax, 0x0050  		; segment
mov    es, ax
mov    bx, [ADDRESS_POSITION]
mov    ah, 0x02    		; read sector(s)
mov    dl, 0x00    		; disk type 
mov    cl, 0x01       		; sector 1
mov    al, 18
pusha
int    0x13
cmp    ah, 0
je     skip_reset
do_reset:
;mov    ax, 0
;delay:
;inc    ax
;cmp    ax, 0x100		; some delay
;jne    delay
mov    ah, 0x00    		; reset drive
mov    dl, 0x00    		; disk type 
xor    ax, ax
int    0x13
;cmp    ah, 0
;jne    do_reset	
jmp    0:next_head			; read again the head
skip_reset:
add    [ADDRESS_POSITION], word (512*18)
cmp    bx, [KERNEL_SIZE]
jg     end_copy 
print_number 	3,2,[ADDRESS_POSITION]
print_number 	4,2,[KERNEL_SIZE]
jmp 0:next_head
end_copy:
;====================================================================

; TODO : Read a Sector for Protection in the future
; Password protection ?
;====================================================================

mov ax, 0xB800
mov es, ax
;;; RM_Write_Boot_Mensage
mov di, 160*0x02 + 2*0x01
mov si, boot_msg
mov ah, 0x03
next_char1:
mov al, byte [si]
mov word [es:di], ax
times 2 inc di
inc si
cmp byte [si],0
jne next_char1

check_keyboard:
in   al, 0x60                
cmp  al, 0x01     	; check ESC key           
jne  check_keyboard          
;mov ah,0	; await key pressed
;int 16h

; Disable Floppy Motor
;mov al,0001100b  ;Disable all motors, controller enabled, DMA  & IRQ enabled
;mov dx,0x3F2     ;Digital Output Register
;out dx,al        ;Sending the command to the controller

; turn off floppy motor
mov dx,3F2h
xor al,al
out dx,al

; JUMP to address 0050:0000 RM (0x00000500 PM)
jmp  0x0050:0


;=======================================================================
boot_msg		DB 	'Loading Boot Sector...',0 
FLOPPY_SIZE		equ	(512*18*2*80)	; 1474560
CLUSTER_SIZE		equ	2048
FS_START		equ	512
FS_ENTRIES              equ	(FLOPPY_SIZE / CLUSTER_SIZE)	; 720
ROOTDIR_START		equ	(FS_START+(4*FS_ENTRIES))+192	; 3584
HEADER_POSITION		equ	ROOTDIR_START+CLUSTER_SIZE+(4*512) ; 7680
HEADER_HEAD             equ     (((HEADER_POSITION/512) / 18) % 2)
HEADER_TRACK            equ     ((HEADER_POSITION/512) / (18*2))
HEADER_SECTOR		equ	((HEADER_POSITION/512) % 18) + 1
KERNEL_POSITION		equ	ROOTDIR_START+CLUSTER_SIZE+(4*512)+CLUSTER_SIZE
KERNEL_HEAD		equ 	(((KERNEL_POSITION/512) / 18) % 2)
KERNEL_TRACK		equ	((KERNEL_POSITION/512) / (18*2))
KERNEL_SECTOR		equ	((KERNEL_POSITION/512) % 18) + 1 
KERNEL_SIZE             DW      0
SECTOR			DB 	0
HEAD			DB	0
TRACK			DB	0
ADDRESS_POSITION	DW	0
NUM_SECTORS		DB	0

;S = (lsn mod spt) + 1
;H = (lsn / spt) mod nh
;T = lsn / (spt*nh)
;
;S is the physical sector number
;H is the head number
;T is the track number
;lsn is the logical sector number
;spt is the number of sectors per track
;nh is the number of heads 
;/ is an integer division
;mod is the modulo operation

;-----------------------------------------------------------------------
times 510-($-$$) DB 0
DW	0xAA55 		; BOOT LOADER SIGNATURE
;-----------------------------------------------------------------------

; PM Address(REAL) = 16 * RM Segment + RM Offset
;  Memory Address   
;     65536        = 16 * 4096 + 0 
;                  = (RM Segment << 4) + RM Offset
;
; 16*0x1000 = 65536*1 start first segment 
; 16*0x2000 = 65536*2 start second segment
; 16*0x3000 = 65536*3 start third segment


heavenos-0.1.4/bootloaders/boot_stdio.asm
[BITS 16]
[ORG 0x7C00]

; DOS header alike boot sector 
;==========================================================
;DB 0xEB,0x1E,0x90 ; SHORT JUMP to 0x1E
;DB 'HeavenOS'     ; OEM Name and Number
;DW 512		  ; Bytes per Sector (512)
;DB 0	 	  ; Sectors per Allocation Unit (cluster)
;DW 0		  ; Reserved seectors (for Boot Record)
;DB 0		  ; Number of FATs
;DW 0		  ; Number of Root Directories Entries
;DW 0		  ; Number of logical Sectors
;DB 0xF0           ; Medium Descriptor Byte (0xF0 - FD, 0xF8 - HD)   
;DW 0		  ; Sectors per FAT
;DW 18		  ; Sectors per TRACK
;DW 2		  ; Number of Heads
;DW 0		  ; Number of Hidden Sectors
;DW 0xAA55	  ; Load Signature - 0xAA55	
;==========================================================
; Boot Loader
;==========================================================

; save EDX value in memory (the last 4 bytes of conventional memory)
; CPU ID Information
mov ax, 0x9000
mov es, ax
mov di, 0xFFFC
mov dword [es:di], edx

; setting regs
mov ax, 0 
mov ds, ax
mov es, ax
mov ss, ax
mov sp, (0x7C00 + 0x400)
mov bp, 0 

jmp  0:start

; COLD REBOOT
cold_boot:
mov             ax, 0x0040                      ; ax = 0040h
mov             es, ax                          ; point es to ax
mov             word [es:0072h], 0x0000         ; Reset
db              0xEA                            ; JUMP
dw              0x0000                          ; FFFF:0000
dw              0xFFFF                          ; ! REBOOTING !
;int     0x19                                   ; warm boot pc 

start:

;=======================================================================
; Check if CPU is 386+
;=======================================================================
pushf                   ; save flags for later
xor     ah,ah           ; clear high byte
push    ax              ; push AX on the stack
popf                    ; pop this value into the flag register
pushf                   ; push flags on the stack
pop     ax              ; ...and get flags into AX
and     ah,0f0h         ; try to set the high nibble
cmp     ah,0f0h         ; on a 80386, the high nibble can never be 0f0h
je      cold_boot
mov     ah,70h          ; now try to set NT and IOPL
push    ax
popf
pushf
pop     ax
and     ah,70h          ; if they couldn't be modified, no 386 is installed
jz      cold_boot
popf                    ; restore flags
;=======================================================================

; set ES to video address
mov ax, 0xB800
mov es, ax

;;; RM_Clear_Screen
mov cx, 0x0fa0
mov di, 0
next_char0:
mov dword [es:di], 0
add di, 4
cmp di, cx
jne next_char0

;;; RM_Write_Boot_Mensage
mov di, 160*0x01 + 2*0x01
mov si, boot_msg
mov ah, 0x03
next_char:
mov al, byte [si]
mov word [es:di], ax
times 2 inc di
inc si
cmp byte [si],0
jne next_char


;=======================================================================
; Copying Kernel Image from Floppy Filesystem to Memory
; point es:bx to 0x0050:0x0000
;=======================================================================
mov    ax, 0x0050  		; segment
mov    es, ax          
mov    bx, 0x0000  		; offset 
; load data from disk to [es:bx]
mov    ah, 0x02    		; read sector(s)
mov    dl, 0x00    		; disk (0x00 - First FD, 0x80 - First HD)
mov    dh, 0       		; head  
mov    ch, 0       		; track (0-1023)
mov    cl, 2      			; sector (2 higher bits = 10-9 bits of track)
mov    al, 17	   		; max. sectors = 18 

int  	0x13			;READ

;second pass
add	bx, 512*17		; next offset
mov dh, 1			; another head  
mov ch, 0			; track 0 (0-1023)

next_sectors:
mov    ax, 0x0050  		; segment
mov    es, ax          
;mov    bx, 0x0000  		; offset 
; load data from disk to [es:bx]
mov    ah, 0x02    		; read sector(s)
mov    dl, 0x00    		; disk (0x00 - First FD, 0x80 - First HD)
mov    cl, 1      			; sector (2 higher bits = 10-9 bit of track)
mov    al, 18	   		; max. sectors = 18  (was 18)

int    0x13

jmp read_out 			; skip the following code it has read enough for 12k
; temp skip this
inc	dh
add	bx, 512*18
cmp dh, 2
jne next_head
mov	dh, 0
inc ch
next_head:
cmp ch, 2
jne next_sectors 

;cmp    ah, 0
;je     skip_reset
;do_reset:
;mov    ah, 0x00    		; reset drive
;mov    dl, 0x00    		; disk type 
;xor    ax, ax
;int    0x13
;jmp short copy_kernel
;skip_reset:

read_out:
;====================================================================
; TODO : Read a Sector for Protection in the future
; Password protection ?
;====================================================================

mov ax, 0xB800
mov es, ax
;;; RM_Write_Boot_Mensage
mov di, 160*0x02 + 2*0x01
mov si, jump_msg
mov ah, 0x03
next_char1:
mov al, byte [si]
mov word [es:di], ax
times 2 inc di
inc si
cmp byte [si],0
jne next_char1

check_keyboard:
in   al, 0x60                
cmp  al, 0x01     	; check ESC key           
jne  check_keyboard          

; Disable Floppy Motor
;mov al,0001100b  ;Disable all motors, controller enabled, DMA  & IRQ enabled
;mov dx,0x3F2     ;Digital Output Register
;out dx,al        ;Sending the command to the controller

; turn off floppy motor
;mov dx,3F2h
;xor al,al
;out dx,al

; JUMP to address 0050:0000 RM (0x00000500 PM)
jmp  0x0050:0


;=======================================================================
boot_msg		DB 	'Loading Boot Sector...',0 
jump_msg		DB	'Press ESC for Jumping to Kernel 32-bit PM Code...',0

debug_msg		DB	'Here',0

Yval			DW	2   ; current Y console position

;============================
; added by stdio	START
;============================

print_d:
	push es			; Save es register
	pusha			; save general purpose registers (E)AX, (E)CX, (E)DX, (E)BX, (E)SP, (E)BP, (E)SI, (E)DI
	; increment Line on print
	mov si,Yval		; Address of Yval
	mov di, word [si]	; copy contents pointed by si (Yval)
	add di,1		; sum + 1
	mov word [si] ,di	; store in mem pointed by si (Yval)
	imul di,160 		; Y*columns (160)
	add di,2 		; X value

	; copy char to video , text mode	
	;mov di, 160*2 + 2*0x01
	mov ax,0xB800		; Address of text memory
	mov es,ax		; Make es, 0xB800
	mov si, debug_msg	; si is address of debug_msg
	mov ah, 0x04		; color/attribute
	pd_next_char:
	mov al, byte [si]	; char
	mov word [es:di], ax	; copy char to video mem
	times 2 inc di		; increase by 2 video
	inc si			; point to next char
	cmp byte [si],0		; is end of string?
	jne pd_next_char	; if not go to next char
	popa			; restore general purpose registers
	pop es			; restore es
ret
;======================== END

;S = (lsn mod spt) + 1
;H = (lsn / spt) mod nh
;T = lsn / (spt*nh)
;
;S is the physical sector number
;H is the head number
;T is the track number
;lsn is the logical sector number
;spt is the number of sectors per track
;nh is the number of heads 
;/ is an integer division
;mod is the modulo operation

;-----------------------------------------------------------------------
times 510-($-$$) DB 0x0 ; stdio, nop operation
DW	0xAA55 		; BOOT LOADER SIGNATURE
;-----------------------------------------------------------------------

; PM Address(REAL) = 16 * RM Segment + RM Offset
;  Memory Address   
;     65536        = 16 * 4096 + 0 
;                  = (RM Segment << 4) + RM Offset
;
; 16*0x1000 = 65536*1 start first segment 
; 16*0x2000 = 65536*2 start second segment
; 16*0x3000 = 65536*3 start third segment
